### DISCUSSION TITLE
[Feature discussion] Allow keeping some elements mounted offscreen

### DISCUSSION THREAD

**Discussion by khovansky-al** [petyosi/react-virtuoso#584] [Ideas]
I’m working on integrating virtuoso into our chat history. We have a feature where text selection is synced across all participants in the room. Whenever a currently selected element is unmounted, the selection state is lost. FF fires a `selectionchange` event. Chrome and Safari don’t fire an event but the selection object becomes mangled.

The only reliable way to make this work with virtualization is by leaving the highlighted elements mounted when they go offscreen. If the key and the order of the persisted elements remains stable, React will not touch them, preserving selection.

I came up with a workaround that gives `Virtuoso` custom `List` and `Item` components and modifies `children` directly.
It looks something like this.

```typescript
// Excerpt from `children` wrapper. This is a component that goes inside `List` (parts omitted for brevity)

// `children` is the original children array passed down by Virtuoso
// child index is taken from data-item-index
const firstRenderedChildIndex = getChildIndex(children[0]);
const lastRenderedChildIndex = getChildIndex(children[children.length - 1]);

// "parked" is how we call children that are invisibly rendered offscreen
const itemsParkedAboveViewport = parkedItemsIndexes
 .filter((index) => index < firstRenderedChildIndex)
 .map((index) => {
   return getParkedItem(index, context);
 });

const itemsParkedBelowViewport = <...same idea...>

return (<>
 {[
   ...itemsParkedAboveViewport,
   ...children,
   ...itemsParkedBelowViewport,
 ]}
</>)

const getParkedItem = (
   index: number,
   context: IVirtualRowContext,
 ) => {
   // get item from the context here

   // ChatVirtualListItem is the same component as we pass to Virtuoso to serve as Item
   // parked prop adds special styling that visually hides the element, moves it out of the way and makes it ignore pointer events
   return (
     <ChatVirtualListItem
       parked
       key={getVirtualRowKey(index, item, context)}
       data-index={index}
       data-item-index={index}
       data-known-size={1}
     >
       {renderChatVirtualItemContent(index, item, context)}
     </ChatVirtualListItem>
   );
 }
```

By keeping the component and the key same, React doesn’t move or touch the components that go offscreen in any way, preserving selection.
This is obviously pretty hacky. For example, Virtuoso complained about 0-sized elements so I made them 1px, and force data- props on them.
I don’t know if there is more child introspection happening internally which could be subtly broken by injecting children like this.
Would really appreciate if you could tell whether there is anything else to watch out.

Considering the case described above, would you be interested in having this officially supported by the library?
I am happy to make a PR but figured something like this would require discussing the API, scope and specifics of integration.
I prototyped this locally, adding a `parkedItems` prop to `Virtuoso` List mode and got children rendering in a similar fashion, but it was of course a quick demo and fully fleshed out implementation would need more thought on how to wire it with all the systems of emitters.

It’s a bit different from top items feature or footer items request seen in a different issue. In my proposal the items have to stay precisely within the “main” item collection to ensure React never moves the corresponding DOM nodes around.
Moving the DOM, even if it’s not explicitly removed but just reparented still causes loss of selection.

Let me know if you need more examples or an interactive demo to better understand what I’m trying to achieve here.

Thank you.

**Comment by petyosi** [comment]
Hey,

I'm not sure that I understand the feature in your chat (text selection being synced), but I am not your PM nor your users, so there's not much of a point to drill further.

To start with your question, I don't think that there's a realistic path towards making such a feature part of the library, especially in the shape you've outlined. People will use it for the wrong reasons. Let's see why:

First, you're probably aware of this yourself; using the DOM as a stateful source of truth for the state in React (especially with virtualization, but not only that) is a path of pain and bugs. I would focus on finding a way to "fix" that, perhaps by emulating a selection that can be kept in React state.  

Next, as you have already guessed, what you're doing is not good for the Virtuoso internals, causes a huge performance hit, and, unless I am missing something, pointless. You're disabling the virtualization - why do you use virtualized components at all? The internet is full of tutorials for infinite, non-virtualized React lists. Maybe I don't see the full picture here, though. 

Finally, here's an idea on how to hack the component more gracefully - use the `increaseViewportBy` property with some very high numbers. You can even manage those dynamically, extending the "rendered" zone based on some logic. 

Hope this helps. My day job is also doing chats. It's one of those surprisingly complex domains - good luck with it.

**Reply by khovansky-al** [reply]
Thanks for the response.
Looks like I did a poor job explaining the situation and trying to do so using our feature only made things more confusing.
My apologies. Allow me to try again with a more common example.

Let’s say we want to preserve native selection, no extra whistles.
A user highlights some text in a message, then scrolls the message out of view far enough that it would unmount thanks to virtualization. We don’t want to lose said selection.

Selection can be observed using [Selection API](https://developer.mozilla.org/en-US/docs/Web/API/Selection)’s `selectionchange` event and stored in React state. Using this interface for restoration, however, would carry a lot of additional complexity and would likely be unreliable, if at all feasible.

`selectionchange` event doesn’t tell you whether the change was triggered by the user or as a consequence of the DOM change.
As mentioned in the original post, Chrome and Safari never fire an event when selection changes due to DOM modifications. They quietly mutate the selection object (`window.getSelection()` result) instead.
You’d need extra heuristics in Firefox to ignore selection events caused by unmounting while preserving intentional events where a user clicks away to clear, etc. This has the potential to race with actual user input and would likely never work as intended.
Of course, there could be many more edge cases to handle.

Selection could span several nodes. Part of the selection could be scrolled off-screen as you select more content. Trying it on the [Virtual List With 100,000 Items](https://virtuoso.dev/hello/) example from Virtuoso website, the parts that go off-screen get de-selected. In this case, saving/restoring parts of the selection wouldn’t help. It also probably serves as a prime example of a situation where you could expect a lot of racing between nodes unmounting and new selection happening.

An alternative that doesn’t require trying to reimplement native selection is to keep the selected node(s) around.
[react-virtual](https://github.com/TanStack/react-virtual), for example, accepts a prop called `rangeExtractor` ([API reference](https://react-virtual.tanstack.com/docs/api)) which allows you to specify which elements of the data collection should be rendered.

If you have 100000 chat messages, you could be rendering, say, messages 30-60 that are currently within the viewport + message number 10 that the user selected but scrolled out of view. You’re not disabling virtualization at all.
If a re-render happens, you still only re-render those visible ones + the extra one kept around.

This could be misused to kill the benefits of virtualization if you put too many extra indexes there, but I suppose the same could be said about overscan, for example.

Hope this clears things up a bit.

P.S. Here's a visual demo.

#### Selecting a long thread in Slack. Slack also uses virtualization.
<video src=https://user-images.githubusercontent.com/4955642/155129209-cf8582e8-5048-45cf-b477-3f22133cd6a7.mov></video>

#### Selecting several items in Virtuoso.
<video src=https://user-images.githubusercontent.com/4955642/155129338-d8ea0895-310a-4fac-8a91-64dfe951c68c.mov></video>

**Reply by petyosi** [reply]
I see what you mean and I do understand that for your use case selection preservation can be an important requirement. 

Preserving certain elements is, indeed, doable, but not trivial. I'm hesitant to mislead you towards promising that I can accept such a PR in the library, as this would need to touch some of the core aspects of [the list state](https://github.com/petyosi/react-virtuoso/blob/master/src/listStateSystem.ts) and the item measurements (the most obvious approach would be to put those preserved items offscreen and skip measuring them). At this point, I am quite conservative about introducing potentially breaking behaviors there.

But I don't want to discourage you from trying, as long as you don't expect me to eventually accept a PR :). Why don't you fork and try an implementation? You might come up with something simpler and more robust than I am imagining. I will take a look.

**Reply by hipstersmoothie** [reply]
Would love this feature too!

my use case: a virtualized drag and droppable list

**Reply by petyosi** [reply]
@hipstersmoothie Great! Are you planning on contributing it? I did not hear back from the original author.

**Reply by khovansky-al** [reply]
For the record, I ended up implementing this feature in userland. It's slightly complicated, but a little over 1.5 years later it's still going strong.
Didn't have time to attempt something that has a 95% chance of being rejected, seeing how our initial conversation went.

**Reply by Darren120** [reply]
@khovansky-al whats userland? can i get the github repo?
